albumentations/augmentations/transforms.py

```diff
@@ -878,6 +878,89 @@ def get_transform_init_args_names(self):
         return ('height', 'width', 'erosion_rate', 'interpolation')
 
 
+class CropNonEmptyMaskIfExists(DualTransform):
+    """Crop area with mask if mask is non-empty, else make random crop.
+
+    Args:
+        crop_height (int): vertical size of crop in pixels
+        crop_width (int): horizontal size of crop in pixels
+        ignore_values (list of int): values to ignore in mask, `0` values are always ignored
+            (e.g. if background value is 5 set `ignore_values=[5]` to ignore)
+        ignore_channels (list of int): channels to ignore in mask
+            (e.g. if background is a first channel set `ignore_channels=[0]` to ignore)
+        p (float): probability of applying the transform. Default: 1.0.
+
+    Targets:
+        image, mask
+
+    Image types:
+        uint8, float32
+    """
+
+    def __init__(self, height, width, ignore_values=None,
+                 ignore_channels=None, always_apply=True, p=1.):
+        super(CropNonEmptyMaskIfExists, self).__init__(always_apply, p)
+
+        if ignore_values is not None and not isinstance(ignore_values, list):
+            raise ValueError('Expected `ignore_values` of type `list`, got `{}`'.format(type(ignore_values)))
+        if ignore_channels is not None and not isinstance(ignore_channels, list):
+            raise ValueError('Expected `ignore_channels` of type `list`, got `{}`'.format(type(ignore_channels)))
+
+        self.height = height
+        self.width = width
+        self.ignore_values = ignore_values
+        self.ignore_channels = ignore_channels
+
+    def apply(self, img, x_min=0, x_max=0, y_min=0, y_max=0, **params):
+        return F.crop(img, x_min, y_min, x_max, y_max)
+
+    @property
+    def targets_as_params(self):
+        return ['mask']
+
+    def get_params_dependent_on_targets(self, params):
+        mask = params['mask']
+        h, w = mask.shape[:2]
+
+        if self.ignore_values is not None:
+            ignore_values_np = np.array(self.ignore_values)
+            mask = np.where(np.isin(mask, ignore_values_np), 0, mask)
+
+        if mask.ndim == 3 and self.ignore_channels is not None:
+            target_channels = np.array([ch for ch in range(mask.shape[-1])
+                                        if ch not in self.ignore_channels])
+            mask = np.take(mask, target_channels, axis=-1)
+
+        if self.height > h or self.width > w:
+            raise ValueError('Crop size ({},{}) is larger than image ({},{})'.format(
+                self.height, self.width, h, w))
+
+        if mask.sum() == 0:
+            x_min = random.randint(0, w - self.width + 1)
```

Comment:
`np.randint(0, N)`  outputs integers from `0` to `N`, including `N`.

=>

`w = 100`
`self.width = 10`

may generate for `x_min` numbers from `0` to `100 - 10 + 1 = 91`, while maximum number to start the crop from is `99 - 10 = 89`

Is it possible, that it should be `-1` and not `+1` ?

Reply:
Thanks, should be '0' actually, I have switched from `np.random` to `random` module to pass your test with serialization (because only random.seed is fixed), and they have different behaviour:
 - np.randon.randint -> range [a, b)
 - randon.randint -> range [a, b]